[proposal] Improve taggers (#2804)
GoogleContainerTools/skaffold: docs/design_proposals/digest-tagger.md

# Improve taggers

* Author(s): David Gageot (@dgageot)
* Date: 4 September 2019

## Background

So far, Skaffold supports multiple taggers or tag policies:

 + the `git` tagger will use git commits/references to tag images
 + the `sha256` tagger always uses `latest` or the tag specified on the artifact's image name
 + the `envTemplate` tagger that can use environment variables to tag images

> The default tagger, is none is specified in the skaffold.yaml, is the git tagger.

Reviewer: ```suggestion
The default tagger, if none is specified in the skaffold.yaml, is the `git` tagger.
```

There are multiple issues:

 + the `git` tagger requires users to have the `git` binary installed. It also requires
   the project to be a git project, which is typically not the case when users just try
   to get started.
 + the `sha256` has a wrong name. It is named like that because, in the end, when Skaffold
   deploys to a remote cluster, the image's sha256 digest will be used as the immutable tag.
   Users are very confused with this name and behaviour.
 + the `sha256` used to be able to use the image tags provided in the artifact definition,
   instead of `latest`. This was not documented and is not possible anymore because artifact
   definitions are now considered invalid if images names have a tag.
 + the `envTemplate` tagger used to be able to replace `{{.DIGEST}}` with the image's imageID
   or digest. This was buggy and is not possible anymore since tags are computed before the
   images are built.
 + users want a tagger that uses the inputs' digest as a tag. They think that's what the
   `sha256` tagger should do.
   
## Proposal

>  + we introduce a `latest` tagger tag tags images with `:latest`.

Reviewer: This is basically a current `sha256`, right? I.e it would still use the `@sha256:sdfsdfsdf` digest at deplyoment time? Or just pure `myimage:latest`? 
How does that behave when the use has an image that already has a tag? overrides the tag? 
Can you elaborate these here?

Author: See below: "No matter the tagger, Skaffold will keep on using immutable references in manifests"

>  + the `latest` tagger is used by default instead of the `git` tagger.

Reviewer: +1

>  + `sha256` is completely changed to use a digest of the artifact's inputs as the tag.

Reviewer: and it won't use the Docker image digest then?

   Something like https://github.com/GoogleContainerTools/skaffold/pull/2301
 + `envTemplate` learns how to replace `{{.DIGEST}}` with a digest of the artifact's
    inputs as the tag.

## Open Issues/Questions

>  + How do we handle users who didn't configure a tagger and were happy with the default

Reviewer: We can introduce a new version - `skaffold fix` should include git tagger between the two versions. this is crappy if there is no actual config change...maybe we can wait until there is a config change? I think we might have to consider `default` value changes as config change.

   being the `git` tagger?

>  + How do we handle users who were happy with `sha256` tagger?

Reviewer: Same as above - `skaffold fix` would migrate `sha256` to `latest`? 
+ for all the above changes we need clear communication in the release notes.

Reviewer 2: as long as there's clear messaging about the name change (and why we did it) then I think this is fine, since the end result will be the same.

# Improve taggers

* Author(s): David Gageot (@dgageot)
* Date: 4 September 2019

## Background

So far, Skaffold supports multiple taggers or tag policies:

 + the `git` tagger uses git commits/references to tag images.
 + the `sha256` tagger uses `latest` or the tag specified on the artifact's image name.
 + the `envTemplate` tagger uses environment variables to tag images.
 + the `datetime` tagger uses current date and time, with a configuration pattern.

The default tagger, if none is specified in the `skaffold.yaml`, is the `git` tagger.

Here are some rules about how tagging currently works:

 + Image tags are computed before the images are built.
 + In early versions of skaffold, we tried to compute tags after the build. It made the process
   super complex with a lot of retagging. It also produced images that were tagged with their
   own digest or imageID which is superfluous since the digest/imageID can be used to reference
   the images directly.

>  + No matter the tagger, Skaffold always uses immutable referenced in Kubenetes manifests.

Reviewer 3: ```suggestion
 + No matter the tagger, Skaffold always uses immutable references in Kubenetes manifests.
```

>    Which reference is used depend on whether the images are pushed or not:

Reviewer 2: ```suggestion
   Which reference is used depends on whether the images are pushed or not:
```

>      + When images are pushed, their immutable digests are available. So images are referenced

Reviewer 2: ```suggestion
     + When images are pushed, their immutable digests are available - this means that images are referenced
```

       by image tag and digest. Something like `image:tag@sha256:abacabac...`. Using both the
       tag and the digest seems superfluous but it guarantees that Kubernetes uses the exact image
       version, referenced by its immutable digest while helping the user quickly see which
       version of the image is used.

>      + When images are NOT pushed, digests are not available. All we have is the tags

Reviewer 3: An example here is would also be useful for comparison like `image:tag`

       and imageIDs. Since imageIDs can't be used in Kubernetes manifests, Skaffold creates
       an additional tag with the same name as the imageID and uses that in manifests.
       This ensures that the exact image version is used.
 + We can never reference images just by their tags because those tags are mutable and
   can lead to cases where Kubernetes will use an outdated version of the image.

## Issues

 + the `git` tagger requires users to have the `git` binary installed. It also requires
   the project to be a git project, which is typically not the case when users just try
   to get started.

>  + the `sha256` has a wrong name. It is named like that because, in the end, when Skaffold

Reviewer 2: ```suggestion
 + `sha256` is a misleading name. It is named like that because, in the end, when Skaffold
```

   deploys to a remote cluster, the image's sha256 digest will be used as the immutable tag.
   Users are very confused with this name and behaviour.
 + the `sha256` used to be able to use the image tags provided in the artifact definition,
   instead of `latest`. This was not documented and is not possible anymore because artifact
   definitions are now considered invalid if images names have a tag.
 + the `envTemplate` tagger used to be able to replace `{{.DIGEST}}` with the image's imageID
   or digest. This was buggy and is not possible anymore since tags are computed before the
   images are built.
 + users want a tagger that uses the inputs' digest as a tag. They think that's what the
   `sha256` tagger should do.
   
## Proposal

 + we introduce a `latest` tagger tag tags images with `:latest`.
 + the `latest` tagger is used by default instead of the `git` tagger.
 + `sha256` tagger is deprecated.

>  + An `inputDigest` is added. It uses the digest of the artifact's inputs as the tag.

Reviewer 3: Isnt this duplicating what docker does ?

Reviewer 3: except skaffold does it before the build and docker does it during build.

Reviewer: there is a big difference: docker calculates image content digests we calculate input source digests. The two can differ in case there are external dependencies / sources of randomness or other reasons the build is not deterministic.

Reviewer 3: ok.

 + we introduce a `latest` tagger tag tags images with `:latest`.
 + the `latest` tagger is used by default instead of the `git` tagger.
 + `sha256` is completely changed to use a digest of the artifact's inputs as the tag.

>    Something like https://github.com/GoogleContainerTools/skaffold/pull/2301

Reviewer 4: For the `sha256` digest, I believe it'd make the most sense to derive it from the resulting image (therefore tagging after the build). Otherwise, unless the Skaffold checksum implementation completely reproduces the Docker's one, there always be missing things, or on the contrary, new tags for the unchanged docker image. Examples:

1. https://github.com/GoogleContainerTools/skaffold/pull/2301 hashes `Dockerfile` and `skaffold.yaml`. Therefore, any formatting changes in those files will affect the checksum but won't affect the image content.
2. The implementation above hashes the working directory (which I assume is the directory containing the `skaffold.yaml` file), however, the resulting image may contain files and other build artifacts from the outside. Therefore, the changes outside of the working directory will affect the image content but won't affect the tag.

Reviewer: FWIW #2301 was not merged. I don't think we should hash Dockerfile and skaffold.yaml into the image. 
To your second point: given the same exact commit, an image built twice with changes in external/remote dependencies will result in different sha256 digests given by the registry / image ID given by the docker daemon. Skaffold uses that and is rendered into the fully qualified image names in _kubernetes_, e.g. a `git` tagger based image has the git based tag + the digest: `gcr.io/k8s-skaffold/gke-loadbalancer:v0.37.1-136-g174ea947a@sha256:7a3419895994a6a9e0c251fa5b3e32134aff987c3e0ec046bf81398c48e07cb4`

Having this digest as a _tag_ as well seems superfluous. 

The tag can be the same - like in the case of `latest`, or `envTemplate` or `dateTime` (if the same image is built the exact same moment) - but the image digest will be different. 

Now, we might still have an issue with the naming - `sha256` seemingly generates expectations - how about `depencyDigest` / `inputDigest` / `localChecksum` ?

Reviewer 4: > FWIW #2301 was not merged. I don't think we should hash Dockerfile and skaffold.yaml into the image.

Whether it was merged or not is irrelevant. I just pointed out the downsides of the proposed solution.

> Skaffold uses that and is rendered into the fully qualified image names in _kubernetes_, e.g. a `git` tagger based image has the git based tag + the digest: `gcr.io/k8s-skaffold/gke-loadbalancer:v0.37.1-136-g174ea947a@sha256:7a3419895994a6a9e0c251fa5b3e32134aff987c3e0ec046bf81398c48e07cb4`
> 
> Having this digest as a _tag_ as well seems superfluous.

This is indeed superfluous. If we have the tag which fully identifies the image contents, nothing else is needed by definition. I'm looking for something like `gcr.io/k8s-skaffold/gke-loadbalancer:7a3419895994a6a9e0c251fa5b3e32134aff987c3e0ec046bf81398c48e07cb4` which is how Skaffold v0.21.1 works.

> The tag can be the same - like in the case of latest, or envTemplate or dateTime (if the same image is built the exact same moment) - but the image digest will be different.

This is the part I don't get. Are we talking about docker image tags or some other tags? As far as I understand, multiple images with different digests cannot have the same name and tags. Even if they could, how would you reference a specific image if more than one have the same name and tag?

> Now, we might still have an issue with the naming - `sha256` seemingly generates expectations - how about `depencyDigest` / `inputDigest` / `localChecksum` ?

What I believe would be ideal is `imageDigest` derived directly from the image after it's built. Please see https://github.com/GoogleContainerTools/skaffold/issues/2713 for the details.

Reviewer: > > The tag can be the same - like in the case of latest, or envTemplate or dateTime (if the same image is built the exact same moment) - but the image digest will be different.

> This is the part I don't get. Are we talking about docker image tags or some other tags? As far as I understand, multiple images with different digests cannot have the same name and tags. Even if they could, how would you reference a specific image if more than one have the same name and tag?

Sorry for the confusion, that really didn't make sense how I put it. I guess I was thinking that at deployment time `image:latest@sha256:digest1` will work. When the next version comes, reusing the same tag `image:latest@sha256:digest2` can still work - as tags _are_ mutable. However as soon as the second version is built, version1 of the image cannot be recovered by tag alone, you will need the digest. 

It seems though that one can refer to an image without the tag...using only the digest e.g.: 
`gcr.io/k8s-skaffold/skaffold@sha256:053a87481bb6a7fbed0381809e369463cba893e38acc2e6677e2b322e0d7819d` ... 

in fact, weirdly you can put _any_ tag - if the digest is present, it will have precedence, the registry will use that: 

`docker pull gcr.io/k8s-skaffold/skaffold:blablabla@sha256:2de99f6219f6c51432843dd808cb02623f70bb4ea1e84cc56dd73aa38ff9e410` works as well ... although obviously `blablabla` tag doesn't exist, not even not pointing to the given digest...

Reviewer: OK, I read a bit more to understand the use cases that motivate `sha256` tagger:

@morozov's use case: can't use latest + need a content dependent tag (repo + potential external contents)

From https://github.com/GoogleContainerTools/skaffold/issues/2713#issuecomment-525412192: 
> The use case is that when multiple images are built from a repository, I need images to be tagged the same (sha256) if their content is not affected by a commit. Can this be achieved after reworking the tagger in #1482?

Mainly motivated by:

> `FATA[0007] build failed: build failed: building [quay.io/redacted/redacted]: build artifact: writing image "quay.io/redacted/redacted:latest": MANIFEST_INVALID: manifest invalid; map[message:manifest schema version not supported]`
> As you can see, the image is tagged as latest and is rejected by the repository versioning policy.

@elisiano's use case: prefixing implemented with envTemplate + `{{.DIGEST}}` 
https://github.com/GoogleContainerTools/skaffold/issues/2162#issuecomment-504312366

Do we care about the actual content, or the digest of the input is sufficient? 
If we care about the content digest (including external effects in non-deterministic builds, that don't only depend on the input sources in the repo) that will have to go back to "post build tagging" model - instead of "calculate the tag, and ask the Builder to build&tag".

Reviewer 4: > It seems though that one can refer to an image without the tag...using only the digest e.g.:
> `gcr.io/k8s-skaffold/skaffold@sha256:053a87481bb6a7fbed0381809e369463cba893e38acc2e6677e2b322e0d7819d`

I didn't know about that and _probably_ it would solve my case w/o having tags at all.

> @morozov's use case: can't use latest + need a content dependent tag (repo + potential external contents)

I need a way to address an image by its content as it was previously done via the `sha256` tagger. Probably, the same can be done via the digest instead but there needs to be a way to pass the digest to the `kustomize` (in my case) deployer. Is it possible to do that?

> @elisiano's use case: prefixing implemented with envTemplate + `{{.DIGEST}}`
> [#2162 (comment)](https://github.com/GoogleContainerTools/skaffold/issues/2162#issuecomment-504312366)

As far as I understand his case, he needs to have images content-addressable (same as in my case) but also be able to provide additional fixed "label" to the image to be able to store multiple different images in the same repository. It can be achieved by using fixed tags and `sha256` digests. E.g.:
  - `quay.io/somecompany/experimental:project1@sha256:$DIGEST`
  - `quay.io/somecompany/experimental:project2@sha256:$DIGEST`
  - and so on

It also requires being able to pass the digest to the deployer, same as in my case.

Reviewer 5: Yes, ideally the images would be content-addressable (which would mean post-build tagging).
And I assume that if skaffold can pass the right tag to kustomize, the described format would be acceptable. ie:
* `quay.io/somecompany/experimental:project1@sha256:$DIGEST`

Reviewer 4: Hm… the problem is that only the `project1` piece is the tag. The rest is just an extra condition to reference the image. Which means that `project1@sha256:$DIGEST1` and `project1@sha256:$DIGEST2` cannot co-exist since a tag points to one and only one image.

Reviewer 4: Given the above, can we tag the image with its digest and allow a custom format which could contain the digest?

Author: @morozov Docker images that are pushed to a registry can always be referenced by their digest.

For example, the `alpine:3.10`, as it is today, can be referenced by:
  + Their digest: `alpine@sha256:72c42ed48c3a2db31b7dafe17d275b634664a708d901ec9fd57b1529280f01fb`
 + Tag and digest: `alpine:3.10@sha256:72c42ed48c3a2db31b7dafe17d275b634664a708d901ec9fd57b1529280f01fb`. It is nicer because it's both immutable and useful for the user who can read the version easily.

If you want to reference images by something that's immutable and content addressable, you should use any of those two references. Skaffold uses the second form whenever possible (that is when images are pushed)

You are asking Skaffold to tag by digest. It would produce references that look like that:
`alpine:72c42ed48c3a2db31b7dafe17d275b634664a708d901ec9fd57b1529280f01fb@sha256:72c42ed48c3a2db31b7dafe17d275b634664a708d901ec9fd57b1529280f01fb`. That seems really superfluous. It is redundant and doesn't offer any meaningful information to a human. (Was `72c42ed48c3a2db31b7dafe17d275b634664a708d901ec9fd57b1529280f01fb` produces before `7746df395af22f04212cd25a92c1d6dbc5a06a0ca9579a229ef43008d4d1302a`?)

wdyt? Could you try again explaining why you'd still need to tag images by their digest? I might me missing something important

Reviewer 4: @dgageot all the above makes sense with the only exception described in https://github.com/GoogleContainerTools/skaffold/pull/2804#discussion_r323324809. It's sort of a corner case and there may be other solutions to it.

To recap: we have to sets of repositories:
1. Production repositories (one per image), where the images are referenced as:
   - `quay.io/company/image1@sha256:$DIGEST1`
   - `quay.io/company/image2@sha256:$DIGEST2`

   These images don't require extra tagging and can be referenced by their `sha256` digest.
2. Experimental repository (just for dev deployments):
   - `quay.io/company/experimental:image1@sha256:$DIGEST11`
   - `quay.io/company/experimental:image1@sha256:$DIGEST12`
   - `quay.io/company/experimental:image2@sha256:$DIGEST21`
   - `quay.io/company/experimental:image2@sha256:$DIGEST22`

   Right now, the images are referenced by their tags generated by `skaffold`. The tags contain the image name (for identification purposes) and the digest (to be content-addressable). This will not work since one tag cannot reference more than one disget.

The question is, how do we store those experimental/dev/ephemeral images in one repository while having them content-addressable and identifiable by image name.

Author: Thank you @morozov for the clarifications. It really helps.

For what it's worth, when using the full form `image:tag@sha256:digest`, the tag is ignored so you can have both `quay.io/company/experimental:image1@sha256:$DIGEST_1` and `quay.io/company/experimental:image1@sha256:$DIGEST_2`. Doesn't really make sense but the correct images will still be used.

In your case, images will always be referenced by a long identifier, any of the following: 
 1. `quay.io/company/experimental:image1-UNIQUE-ID`
 2. `quay.io/company/experimental:image1-DIGEST`
 3. `quay.io/company/experimental:image1-UNIQUE-ID@sha256:DIGEST`
 4. `quay.io/company/experimental:image1-DIGEST@sha256:DIGEST`
 5. `quay.io/company/experimental:image1@sha256:DIGEST` (`image1` is ignored but still gives a hint to the user that they are using the right image)
 6. `quay.io/company/experimental@sha256:DIGEST`.

The last two options are possible with both the current state of Skaffold and the proposed state. And they look ok to me when I compare them with the other options. wdyt?

I understand you'd rather have option 2 or 4 but it requires computing tags after the build which is something we want to avoid because we've been there and it's a mess.

Another possibility is use the `inputDigest` as a pseudo unique id. I think it's kinda ok but as you said, there's no bijection between such id and the digest.

Reviewer 4: > For what it's worth, when using the full form `image:tag@sha256:digest`, the tag is ignored so you can have both `quay.io/company/experimental:image1@sha256:$DIGEST_1` and `quay.io/company/experimental:image1@sha256:$DIGEST_2`. Doesn't really make sense but the correct images will still be used.

As far as I understand, it will work like this from the referencing standpoint. But what about the repository? Will skaffold be able to create two images in the same repository with the same tag but different digests?

> I understand you'd rather have option 2 or 4 but it requires computing tags after the build

I don't really have a preference in which syntax I'll end up using. I need a migration path from v0.29 to the latest one. Option #‌2 looks closest to what we have now; #‌4 indeed looks redundant.

If you believe that #‌5 will work, I could give it some more testing but I'll need some extra guidance. What I tried was:
```
$ docker images
REPOSITORY                     TAG                 IMAGE ID            CREATED             SIZE
mariadb                        latest              97df12fa9319        6 months ago        369MB
postgres                       latest              d03e2a8f3ed4        8 months ago        312MB

$ docker tag mariadb:latest experimental:project1

$ docker tag postgres:latest experimental:project1

$ docker images
REPOSITORY                     TAG                 IMAGE ID            CREATED             SIZE
mariadb                        latest              97df12fa9319        6 months ago        369MB
experimental                   project1            d03e2a8f3ed4        8 months ago        312MB
postgres                       latest              d03e2a8f3ed4        8 months ago        312MB
```

As you can see, there's only one `experimental:project1` tag that points to the same image as `postgres:latest`.

Reviewer 4: @elisiano what is the real application of the `project-` prefixes in the experimental repository? Are they used for some automated housekeeping or mostly in order for a human to understand what project a specific image belongs to? In other words, what are the consequences of not using tags in the experimental repository but instead always using the `@sha256` digest?
